pipeline/pipeline/resource.py

```diff
@@ -96,13 +130,63 @@ def _get_path(self, directory):
 
 
 class TaskResourceFile(ResourceFile):
+    """
+    Class representing an intermediate file from a task.
+
+    Notes
+    -----
+    All :class:`.TaskResourceFile` are temporary files and must be written
+    to a permanent location using :func:`.Pipeline.write_output` if the output needs
+    to be saved.
+    """
+
     def _get_path(self, directory):
         assert self._source is not None
         assert self._value is not None
         return directory + '/' + self._source._uid + '/' + self._value
 
 
 class ResourceGroup(Resource):
+    """
+    Class representing a mapping of identifiers to a resource file.
+
+    Examples
+    --------
+
+    Initialize a pipeline and create a new task:
+
+    >>> p = Pipeline()
+    >>> t = p.new_task()
+
+    Read a set of input files as a resource group:
+
+    >>> bfile = p.read_input_group(bed="data/example.bed",
+    ...                            bim="data/example.bim",
+    ...                            fam="data/example.fam")
+
+    Reference the entire file group:
+
+    >>> t.command(f"plink --bfile {bfile} --geno 0.2 --out {t.ofile}")
+
+    Reference a single file:
+
+    >>> t.command(f"wc -l {bfile.fam}")
+
+    Create a resource group from a task intermediate:
+
+    >>> t.declare_resource_group(ofile={'bed': '{root}.bed',
+    ...                                 'bim': '{root}.bim',
+    ...                                 'fam': '{root}.fam'})
+    >>> t.command(f"plink --bfile {bfile} --make-bed --out {t.ofile}")
```

Comment:
Along the lines of the comment above, it's implied earlier that:
```
`dest` is the file root and each resource file will be written to `{root}.identifier`
```
But here you actually declared the extensions?

Reply:
Sorry, this is hard to explain and maybe I need to rethink the interface. `--out` will write a `.bed`, `.bim`, `.fam` file to the path given by `{t.ofile}`. However, I need to capture the fact that there are 3 files generated for file localization purposes in the future. For example, 

```
t1 = p.new_task().command('plink --bfile {input} --make-bed --out {t1.bfile}
t2 = p.new_task().command('plink --bfile {t1.bfile} --make-bed --out {t2.bfile}
```

`t1.bfile` will be rendered as the file root (same passed to `--out` in `t1`). But there are 3 files we need to capture `{root}.bed`, ... where `{root}` is equivalent to the root path stored by `t1.bfile`. The key words `bed`, `bim`, and `fam` could have been anything. What matters is the expression with `{root}` in it matches the output file paths being generated by PLINK. The `bed`, `bim`, `fam` identifiers are so you can reference a specific file from the group. For example, finding duplicate sample ids

```
t3 = p.new_task().command(f'awk \'{{ print $1 }}\' {t2.bfile.fam} | sort | uniq -c > {t3.dups}')
```

The resource group is also necessary for things like the FASTA example here:
https://software.broadinstitute.org/wdl/documentation/article?id=7615

The FASTA file has the actual fasta, an index, and a dict. The `-R` flag only wants the path of the FASTA file. The other two files are implied based on the fasta file path. Pipeline needs a way of knowing that it needs to localize all of the files.

For actually writing the file, I didn't want this
```
p.write_output(t.bfile.bed, "gs://foo.bed")
p.write_output(t.bfile.bim, "gs://foo.bim")
...
```

I thought it was better to do this and auto copy all known associated files:
```
p.write_output(t.bfile, "gs://foo")
```
But I need a way to differentiate the files, so appending the identifier name (which in most cases will be the correct extension) as the extension seemed like a sensible interface. If it's too confusing, I can force only resource files can be written out and then you'd have to do the same thing as above and write each individually. 

If this isn't clear, we can talk in person.